### PR TITLE
Upgrade to `webauthn` 1.6.0

### PR DESCRIPTION
**Pull request by Natim**

### CODE REVIEW

**Comment by dunderrrrrr** [comment]
Is this about to be merged into main when checks pass? @justinmayer 
Would much appreciate webauthn version 1.6.0 :smile:

**Comment by Natim** [comment]
@dunderrrrrr did you had a chance to test kagi with this new branch? I am a bit worried that the migration is not going to be easy.

**Comment by dunderrrrrr** [comment]
@Natim not yet, ill give it a shot when I have time. Thanks!

**Comment by justinmayer** [comment]
> @dunderrrrrr wrote:
> 
> Is this about to be merged into main when checks pass?

I am traveling for the next week, and when I return I will be able to test these changes on a production application.

In the interim, as Rémy suggested, any testing you can do would be much appreciated!

File: kagi/utils/webauthn.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    return generate_challenge()
+
+
+def get_credential_options(user, *, challenge, rp_name, rp_id):
+    """
+    Returns a dictionary of options for credential creation
+    on the client side.
+    """
+    _authenticator_selection = AuthenticatorSelectionCriteria()
+    _authenticator_selection.user_verification = UserVerificationRequirement.DISCOURAGED
```
Comment (inbalzelinger) [inline kagi/utils/webauthn.py:92]: If you want to allow the user to register new `usb` `nfc` yubikey. One that its pin code wasn't defined before,
 you should define the 
`_authenticator_selection.resident_key=ResidentKeyRequirement.PREFERRED` [see](https://github.com/duo-labs/webauthn.io/blob/master/_app/homepage/services/registration.py#L48:~:text=resident_key%3DResidentKeyRequirement.PREFERRED)

**Comment by justinmayer** [comment]
After testing the contents of this `upgrade-webauthn-1.6.0` branch on an existing production application that relocates the Kagi views from its default path, I encountered an error when trying to add a WebAuthn key. I finally found the time to reproduce this error using the bundled test project, which I will document here.

First, **starting on the current `main` branch**, I followed the steps in the README to set up the test project, including running the initial migrations, creating a superuser, etc. Then I logged into `http://localhost:8000` using Firefox 112.0.1, added a new WebAuthn key successfully, logged out, logged back in using the WebAuthn key, and finally removed the WebAuthn key. In short, I confirmed that basic WebAuthn behavior functions correctly in the current `main` branch. In a separate terminal session and virtual environment, I switched to the `upgrade-webauthn-1.6.0` branch, installed the upgraded dependencies and code via `poetry upgrade && poetry install`, ran migrations, and confirmed that all of the above steps continue to function as expected. So far, so good.

Then, switching back to my other terminal session running the current `main` branch, I modified `testproj/testproj/urls.py`, adding these imports…

```python
from kagi.views import api as kagi_api_views
from kagi.views.login import KagiLoginView
```

… these `urlpatterns` …

```python
path("login/", KagiLoginView.as_view(), name="login"),
path("settings/security/", include("kagi.urls", namespace="kagi")),
```

… and I commented out this line, which is obviated by the line above:

```python
# path("kagi/", include("kagi.urls", namespace="kagi")),
```

After making these changes, I confirmed that all of the above WebAuthn-related behavior continues to function as expected on the current `main` branch, even though the Kagi views have been relocated from `/kagi/` to `/settings/security/`.

However, when I switch back to the `upgrade-webauthn-1.6.0` branch, attempting to add a WebAuthn key results in the following error in the browser console:

```
Uncaught (in promise) SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data

    ProvisionWebAuthn http://localhost:8000/static/kagi/webauthn.js:211
    AsyncFunctionThrow self-hosted:814
    (Async: async)
    doWebAuthn http://localhost:8000/static/kagi/webauthn.js:47

GET | http://localhost:8000/kagi/api/begin-activate/
Status: 404 Not Found
```

In short, it seems the new branch depends on the `/kagi/` prefix in a way that the current `main` branch does not, which at least for me feels like a regression.

@Natim: Do you have any ideas about what is going on here, and perhaps how it might be resolved?

(Copying @apollo13, @carltongibson & @MarkusH)

**Comment by justinmayer** [comment]
@rphlo: Around the time of your previous comment, I tried using `window.Kagi.verify_credential_info`, but I could not figure out how to use it to resolve the problem. Is there any chance you could propose a specific diff that would successfully resolve the problem in question?

**Comment by apollo13** [comment]
Hi @rphlo, thank you so much for this. I will pulling your chances into this branch shortly and will start testing.

File: pyproject.toml [hunk 2]
```diff
@@ -50,6 +50,9 @@ pytest-xdist = "^2.1"
 sphinx = "^4.0"
 Werkzeug = "^2.0"
 
+[tool.poetry.group.dev.dependencies]
+pretend = "^1.0.9"
```
Comment (apollo13) [inline pyproject.toml:54]: There seems to be a group `dev-dependencies` already above. I guess those two should get merged.

**Comment by MarkusH** [comment]
> @rphlo: Around the time of your previous comment, I tried using `window.Kagi.verify_credential_info`, but I could not figure out how to use it to resolve the problem. Is there any chance you could propose a specific diff that would successfully resolve the problem in question?

As a follow-up to this, I opened #65

**Review by MarkusH — APPROVED** [review]

File: kagi/static/kagi/webauthn.js [hunk 3]
```diff
@@ -1,329 +1,281 @@
...
-if (typeof window.Kagi === 'undefined') {
-  let Kagi = window.Kagi || {
-    begin_activate: '/kagi/api/begin-activate/',
-    begin_assertion: '/kagi/api/begin-assertion/',
-    verify_credential_info: '/kagi/api/verify-credential-info/',
-    verify_assertion: '/kagi/api/verify-assertion/',
-    keys_list: '/kagi/keys/',
-  };
-  console.error("window.Kagi is not defined, falling back to default URLs", Kagi);
+/* Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (apollo13) [inline kagi/static/kagi/webauthn.js:1]: Where is the Apache license coming all of a sudden? If this code is copied from somewhere else it might make sense to reference it. Otherwise it would be good to stick to one license in the project (which seems to be the simplified 2-clause BSD). @Natim can you shed some light on this?

Comment (justinmayer) [reply kagi/static/kagi/webauthn.js:1]: If memory serves, this came via the excellent work done to implement MFA for the Warehouse project: https://github.com/pypi/warehouse/blob/main/warehouse/static/js/warehouse/utils/webauthn.js

I agree that it makes sense to reference it. 👍

Comment (justinmayer) [reply kagi/static/kagi/webauthn.js:1]: Added via 9f7281a.

Comment (justinmayer) [reply kagi/static/kagi/webauthn.js:1]: Indeed — thanks, Raphaël. I added similar annotations in the missing spots via 4101e13.

**Review by justinmayer — APPROVED** [review]
Many sincere thanks to @Natim for all his work on this endeavor, as well as to everyone else who helped review and improve it! 🌟
